internal/globalratelimit/runner/runner.go

```diff
@@ -109,17 +110,27 @@ func (r *Runner) serveXdsConfigServer(ctx context.Context) {
 	}
 }
 
+func buildXDSResourceFromCache(rateLimitConfigsCache map[string][]cachetype.Resource) types.XdsResources {
+	xdsResourcesToUpdate := types.XdsResources{}
+	for _, xdsR := range rateLimitConfigsCache {
+		xdsResourcesToUpdate[resourcev3.RateLimitConfigType] = append(xdsResourcesToUpdate[resourcev3.RateLimitConfigType], xdsR...)
+	}
+
+	return xdsResourcesToUpdate
+}
+
 func (r *Runner) subscribeAndTranslate(ctx context.Context) {
+	// rateLimitConfigsCache is a cache of the rate limit config, which is keyed by the xdsIR key.
+	rateLimitConfigsCache := map[string][]cachetype.Resource{}
```

Comment:
is there a way to not maintain this cache here and instead create an API like `rCache := r.getSnapshot(ctx)` instead ? this will allow us to reduce one copy in memory

Reply:
Let me try; depending on the complexity that would give though, that looks a better idea.

Reply:
The change would be like this https://github.com/sanposhiho/gateway/commit/a9f0a190ed7ecc939225598c0e5ce96f23307ce9

So, either way, we have to have a cache to track which ir.Key generates which rate limit configurations. (A cache size would be smaller than the current cache of entire configs, that's one good point though.)
The cache is required because otherwise we wouldn't know which part of snapshot we have to update. (A `message.Update` comes from `watchable.Snapshot` doesn't have a previous state of ir.)

So, I'm not very motivated to proceed that way; For me, it looks like only a little gain (less memory usage) would come with a certain downside (complexity).

Reply:
hey @sanposhiho thanks for writing up the quick diff, what I was suggesting was a little different, instead of holding the entire cache on the stack using `rateLimitConfigsCache` which incurs a constant memory hit, can we retrieve the cache snapshot using an API like 
https://github.com/envoyproxy/go-control-plane/blob/1da4500d00e270d803caefbe0c20e4d3d162e586/pkg/cache/v3/snapshot.go#L124
every time we get a new new message ?

Reply:
Sorry, what's the difference from my diff? It fetches the previous snapshot with GetSnapshot -> GetResources, and try to rebuild a new snapshot from there already.

Reply:
I missed that 🙈 
lets go ahead with diff to keep the code complexity low and we can revisit these if/when we hit a memory bottleneck